From: Neil Wilson <neil@brightbox.co.uk>
To: libvir-list@redhat.com
Date: Mon, 10 Jan 2011 09:52:56 +0000
Message-ID: <1294653176.3013.16.camel@lenovo-3000-n100>

Hi,

Here's the patch to add basic ACL support to QEMU within libvirt. Like
SASL it's ignored by RHEL5's default qemu. Newer qemu picks it up as
expected and you can manipulate the acls using 'virsh'. 


diff --git a/src/qemu/qemu.conf b/src/qemu/qemu.conf
index ba41f80..7ab5eee 100644
--- a/src/qemu/qemu.conf
+++ b/src/qemu/qemu.conf
@@ -71,6 +71,15 @@
 # vnc_sasl = 1
 
 
+# Enable the VNC access control lists. When switched on this will
+# initially block all vnc users from accessing the vnc server. To
+# add and remove ids from the ACLs you will need to send the appropriate
+# commands to the qemu monitor as required by your particular version of
+# QEMU. See the QEMU documentation for more details.
+# 
+# vnc_acl = 1
+
+
 # The default SASL configuration file is located in /etc/sasl/
 # When running libvirtd unprivileged, it may be desirable to
 # override the configs in this location. Set this parameter to
diff --git a/src/qemu/qemu_command.c b/src/qemu/qemu_command.c
index 7dd8e03..0ffeaab 100644
--- a/src/qemu/qemu_command.c
+++ b/src/qemu/qemu_command.c
@@ -3542,8 +3542,11 @@ qemuBuildCommandLine(virConnectPtr conn,
                     virCommandAddEnvPair(cmd, "SASL_CONF_DIR",
                                          driver->vncSASLdir);
 
-                /* TODO: Support ACLs later */
             }
+
+	    if (driver->vncACL) 
+	        virBufferAddLit(&opt, ",acl");
+	    
         }
 
         virCommandAddArg(cmd, "-vnc");
--- libvirt-1.0.1/src/qemu/qemu_conf.c.orig	2013-01-08 20:41:45.877258478 +0100
+++ libvirt-1.0.1/src/qemu/qemu_conf.c	2013-01-08 21:27:27.817201325 +0100
@@ -170,6 +170,7 @@
     GET_VALUE_LONG("vnc_sasl", driver->vncSASL);
     GET_VALUE_STR("vnc_sasl_dir", driver->vncSASLdir);
     GET_VALUE_LONG("vnc_allow_host_audio", driver->vncAllowHostAudio);
+    GET_VALUE_LONG("vnc_acl", driver->vncACL);
 
     p = virConfGetValue(conf, "security_driver");
     if (p && p->type == VIR_CONF_LIST) {
diff --git a/src/qemu/qemu_conf.h b/src/qemu/qemu_conf.h
index 83ddedd..3e7506d 100644
--- a/src/qemu/qemu_conf.h
+++ b/src/qemu/qemu_conf.h
@@ -85,6 +85,7 @@ struct qemud_driver {
     unsigned int vncTLS : 1;
     unsigned int vncTLSx509verify : 1;
     unsigned int vncSASL : 1;
+    unsigned int vncACL : 1;
     char *vncTLSx509certdir;
     char *vncListen;
     char *vncPassword;

